File: 207. Course Schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        preq=defaultdict(list)
        for c,p in prerequisites:
            preq[c].append(p)
        cycle=set()
        seen=set()
        def check(course):
            if course in seen:
                return True
            if course in cycle:
                return False
            cycle.add(course)
            for c in preq[course]:
                if not check(c):
                    return False
            cycle.remove(course)
            seen.add(course)
            return True
        for course in range(numCourses):
            if not check(course):
                return False
        return True

File: 207. Course Schedule.cpp
class Solution {
public:
  bool flag = true;
  unordered_map<int, vector<int>> dic;
  unordered_set<int> seen;
  unordered_set<int> cycle;
  void check(int c) {

    if (cycle.count(c) == 1) {
      flag = false;
      return;
    }
    if (flag == false || seen.count(c) == 1)
      return;

    cycle.insert(c);
    for (auto course : dic[c]) {
      check(course);
    }
    cycle.erase(c);
    seen.insert(c);
  }
  bool canFinish(int numCourses, vector<vector<int>> &prerequisites) {

    for (auto pair : prerequisites) {
      dic[pair[0]].push_back(pair[1]);
    }
    for (int c = 0; c < numCourses; c++) {
      if (seen.count(c) == 0)
        check(c);
    }

    return flag;
  }
};
